### PR TITLE
core[minor]: Add graph functionality to Runnable base

### LINKED ISSUE #4582: Port `get_graph` from Python
### Privileged issue

- [X] I am a LangChain maintainer, or was asked directly by a LangChain maintainer to create an issue here.

### Issue Content

This issue focuses on porting `get_graph` implementation from Python, specifically [this function](https://github.com/langchain-ai/langchain/blob/a99eb3abf4db7ce98fb271df93c23867583782b8/libs/core/langchain_core/runnables/base.py#L388-L398).

This is required to build a graph visualizer tool for `langgraphjs`.

### PR DESCRIPTION
**Pull request by albertpurnama**
## Description
- Adds `Graph` class port over from Python's `Graph` implementation [here](https://github.com/langchain-ai/langchain/blob/a99eb3abf4db7ce98fb271df93c23867583782b8/libs/core/langchain_core/runnables/graph.py#L94-L95)
- Adds `getGraph` to base `Runnable` to allow constructing graph representation of the Runnable. This should be exact port of [this function in python](https://github.com/langchain-ai/langchain/blob/a99eb3abf4db7ce98fb271df93c23867583782b8/libs/core/langchain_core/runnables/base.py#L388-L389)

## Failing test case

```
(Use `node --trace-warnings ...` to show where the warning was created)
@langchain/core:test: FAIL src/runnables/tests/runnable_graph.test.ts
@langchain/core:test:   ● Test graph sequence
@langchain/core:test: 
@langchain/core:test:     expect(received).toBe(expected) // Object.is equality
@langchain/core:test: 
@langchain/core:test:     - Expected  - 18
@langchain/core:test:     + Received  +  2
@langchain/core:test: 
@langchain/core:test:     @@ -17,20 +17,11 @@
@langchain/core:test:             "target": 3,
@langchain/core:test:           },
@langchain/core:test:         ],
@langchain/core:test:         "nodes": Array [
@langchain/core:test:           Object {
@langchain/core:test:     -       "data": Object {
@langchain/core:test:     -         "properties": Object {
@langchain/core:test:     -           "name": Object {
@langchain/core:test:     -             "title": "Name",
@langchain/core:test:     -             "type": "string",
@langchain/core:test:     -           },
@langchain/core:test:     -         },
@langchain/core:test:     -         "title": "PromptInput",
@langchain/core:test:     -         "type": "object",
@langchain/core:test:     -       },
@langchain/core:test:     +       "data": Object {},
@langchain/core:test:             "id": 0,
@langchain/core:test:             "type": "schema",
@langchain/core:test:           },
@langchain/core:test:           Object {
@langchain/core:test:             "data": Object {
@langchain/core:test:     @@ -49,11 +40,10 @@
@langchain/core:test:             "data": Object {
@langchain/core:test:               "id": Array [
@langchain/core:test:                 "langchain",
@langchain/core:test:                 "llms",
@langchain/core:test:                 "fake",
@langchain/core:test:     -           "llm",
@langchain/core:test:                 "FakeLLM",
@langchain/core:test:               ],
@langchain/core:test:               "name": "FakeLLM",
@langchain/core:test:             },
@langchain/core:test:             "id": 2,
@langchain/core:test:     @@ -71,17 +61,11 @@
@langchain/core:test:             },
@langchain/core:test:             "id": 3,
@langchain/core:test:             "type": "runnable",
@langchain/core:test:           },
@langchain/core:test:           Object {
@langchain/core:test:     -       "data": Object {
@langchain/core:test:     -         "items": Object {
@langchain/core:test:     -           "type": "string",
@langchain/core:test:     -         },
@langchain/core:test:     -         "title": "CommaSeparatedListOutputParserOutput",
@langchain/core:test:     -         "type": "array",
@langchain/core:test:     -       },
@langchain/core:test:     +       "data": Object {},
@langchain/core:test:             "id": 4,
@langchain/core:test:             "type": "schema",
@langchain/core:test:           },
@langchain/core:test:         ],
@langchain/core:test:       }
@langchain/core:test: 
@langchain/core:test:       33 |   // expect(Object.keys(graph.nodes).length).toBe(5);
@langchain/core:test:       34 |
@langchain/core:test:     > 35 |   expect(graph.toJson()).toBe({
@langchain/core:test:          |                          ^
@langchain/core:test:       36 |     nodes: [
@langchain/core:test:       37 |       {
@langchain/core:test:       38 |         id: 0,
@langchain/core:test: 
@langchain/core:test:       at Object.toBe (src/runnables/tests/runnable_graph.test.ts:35:26)
@langchain/core:test: 
```

This is because the implementation of base `Runnable` is still TBD. Specifically [here ](https://github.com/langchain-ai/langchainjs/pull/4566/files#diff-59ab510ff925ffbaa3959943f899d56160b14d8ffd74ea797ec9aee06fb8ef36R558) (same for the last node)

The solution to this requires the base `Runnable` to be fully implemented

Fixes #4582

### CODE REVIEW

File: langchain-core/src/runnables/base.ts [hunk 1]
```diff
@@ -550,6 +551,29 @@ export abstract class Runnable<
...
   }
 
+  getGraph(config?: RunnableConfig): Graph {
+    const graph = new Graph();
+
+    // TODO: what should the inputNode be?
+    const inputNode = graph.addNode({
+      type: "unknown",
+      config,
+    });
```
Comment (albertpurnama) [inline langchain-core/src/runnables/base.ts:561]: @jacoblee93 Here's where I'm a little bit lost,

I'm trying to find a way to put the schema of the input node of this runnable. I haven't been able to find it yet.

This should be similar to [get_input_schema](https://github.com/langchain-ai/langchain/blob/a99eb3abf4db7ce98fb271df93c23867583782b8/libs/core/langchain_core/runnables/base.py#L393) in Python's core lib. 

I think for graphing purposes, we might be able to replace this with a placeholder and maybe implement this later when we're going to use Zod.

File: langchain-core/src/runnables/base.ts [hunk 2]
```diff
@@ -550,6 +551,29 @@ export abstract class Runnable<
...
+
+    const runnableNode = graph.addNode(this); // Assuming `this` refers to an instance that can be treated as data for a node
+
+    // TODO: what should the outputNode be?
+    const outputNode = graph.addNode({
+      type: "unknown",
+      config,
+    });
```
Comment (albertpurnama) [inline langchain-core/src/runnables/base.ts:569]: Same case as the above.

File: langchain-core/src/runnables/graph.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+interface Edge {
+  source: string;
+  target: string;
+  data?: string;
+}
+
+interface Node {
+  id: string;
+  data: any;
```
Comment (nfcampos) [inline langchain-core/src/runnables/graph.ts:12]: we prob want to type this

File: langchain-core/src/runnables/graph.ts [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+function nodeDataStr(node: Node): string {
+  if (!isUuid(node.id)) {
+    return node.id;
+  } else {
+    // Assuming `node.data` has a similar structure to the Python version
+    let data = node.data.toString();
```
Comment (nfcampos) [inline langchain-core/src/runnables/graph.ts:20]: I think we want to use `node.data.getName()` here

File: langchain-core/src/runnables/base.ts [hunk 5]
```diff
@@ -550,6 +551,22 @@ export abstract class Runnable<
     );
   }
 
+  getGraph(_?: RunnableConfig): Graph {
+    const graph = new Graph();
+
+    // TODO: what is the inputNode here?
+    const inputNode = graph.addNode({} as unknown as Runnable);
```
Comment (albertpurnama) [inline langchain-core/src/runnables/base.ts:558]: @nfcampos can I get some pointers on how to solve this? seems like the `get_input_schema` isn't implemented yet in JS, unlike [in python](https://github.com/langchain-ai/langchain/blob/a99eb3abf4db7ce98fb271df93c23867583782b8/libs/core/langchain_core/runnables/base.py#L393-L394)

**Comment by jacoblee93** [comment]
Hey sorry about the delay - will spend some time with this tomorrow and hopefully ship it!

File: langchain-core/src/runnables/base.ts (hunk 5 above)
Comment (jacoblee93) [reply langchain-core/src/runnables/base.ts:558]: I think it's fine to leave this for now - we can add at a later date when we add support for input/output schemas.

File: langchain-core/src/load/serializable.ts [hunk 6]
```diff
@@ -4,6 +4,9 @@ export interface BaseSerialized<T extends string> {
   lc: number;
   type: T;
   id: string[];
+  name?: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  graph?: Record<string, any>;
```
[the remark this answers is not in the document]
Comment (jacoblee93) [reply langchain-core/src/load/serializable.ts:9]: Yeah I think fine!

File: langchain-core/src/runnables/base.ts [hunk 7]
```diff
@@ -550,6 +551,22 @@ export abstract class Runnable<
...
+
+    const runnableNode = graph.addNode(this); // Assuming `this` refers to an instance that can be treated as data for a node
+
+    // TODO: what is the outputNode here?
+    const outputNode = graph.addNode({} as unknown as Runnable);
```
Comment (jacoblee93) [inline langchain-core/src/runnables/base.ts:563]: It's going to be the equivalent of output schema - we should ideally infer this when we have output schema support.

File: langchain-core/src/runnables/graph.ts [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+  } else if (Runnable.isRunnable(node.data)) {
+    try {
+      let data = node.data.toString();
+      if (
+        data.startsWith("<") ||
+        data[0] !== data[0].toUpperCase() ||
+        data.split("\n").length > 1
+      ) {
+        data = node.data.constructor.name;
+      } else if (data.length > 42) {
```
Comment (jacoblee93) [inline langchain-core/src/runnables/graph.ts:29]: Factor out into a constant?

File: langchain-core/src/runnables/graph.ts [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+  source: string;
+  target: string;
+  data?: string;
+}
+
+interface Node {
+  id: string;
+
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  data: Function | RunnableType;
```
Comment (jacoblee93) [inline langchain-core/src/runnables/graph.ts:14]: We can use `(...args: any[]) => any` I think

File: langchain-core/src/runnables/base.ts [hunk 10]
```diff
@@ -550,6 +551,22 @@ export abstract class Runnable<
     );
   }
 
+  getGraph(_?: RunnableConfig): Graph {
```
Comment (jacoblee93) [inline langchain-core/src/runnables/base.ts:554]: Wouldn't be any reason to preemptively make this async right?

**Comment by jacoblee93** [comment]
Thank you very much for your patience and the great work @albertpurnama!

**Pull request merged**